Tarball

reidmv opened the pull request:
Chunk PR in support of (original monster PR) #1159

This pull request **DEPENDS ON** #1242, #1243. Do not merge until after those PRs are merged.

This PR implements:

* An R10K::Tarball base class
* An R10K::Module::Tarball class, providing a "tarball" module type
* An R10K::Environment::Tarball class, providing a "tarball" environment type

Example usage:

```
---
production:
  type: tarball
  source: https://example.com/puppet-code-env-v1.0.0.tar.gz
  version: 99a906c99c2f144de43f2ae500509a7474ed11c583fb623efa8e5b377a3157f0 # sha256digest
  modules:
    puppetlabs-peadm:
      type: forge
      version: 2.5.0
    reidmv-xampl:
      type: git
      source: https://github.com/reidmv/reidmv-xampl.git
      version: 62d07f2
    reidmv-mymodule:
      type: tarball
      source: https://example.com/reidmv-mymodule-1.0.0.tar.gz
      version: 6128ada158622cd90f8e1360fb7c2c3830a812d1ec26ddf0db7eb16d61b7293f # sha256digest
```

Later enhancements built on top of this might include Artifactory types, or S3 types, with options for real versions, depending on how the remote repository works.

Fixes #567
Closes #1159

reidmv committed:
Implement R10K::Tarball

This class will serve as a base class for creating tarball environment
and module types.

reidmv committed:
Update initializers

Trying to make sure the global cachedir option gets respected by tarball

reidmv committed:
Update r10k.gemspec minitar dependency

It's not just a development dependency; it's an r10k dependency too.
Really it always has been because it's a puppetforge dependency.

reidmv committed:
Implement R10K::Module::Tarball

This permits modules of type tarball to be used. Tarball modules will
be extracted with no modification into the modules target directory.
Tarballs should not be packed to include a top-level containing folder.

    modules:
      lwf-remote_file:
        type: tarball
        source: https://forge.puppet.com/v3/files/lwf-remote_file-1.1.3.tar.gz
        version: bce01b849631926b452900eff5dc385c893c4320c95b290f992e3c529f417f0e

reidmv committed:
Implement R10K::Environment::Tarball

This permits core environment content to be sourced from tarballs. The
tarball environment type supports defining modules as well.

    ---
    production:
      type: tarball
      source: https://example.com/puppet-code-env-1.0.0.tar.gz
      version: 99a906c99c2f144de43f2ae500509a7474ed11c583fb623efa8e5b377a3157f0

Magisus commented on lib/r10k/module/tarball.rb:
```diff
@@ -0,0 +1,95 @@
+  end
+
+  # Synchronize this module with the indicated state.
+  def sync(opts={})
```
Can you make this return whether it was updated or not? Like this does: https://github.com/puppetlabs/r10k/blob/c5628d2063a8a936f083f2e4c724205e4002de4a/lib/r10k/module/git.rb#L100-L111 We use that info to determine whether to run post-deploy hooks on a `deploy module` command.

reidmv replied:
Adjusted return value in c95c9c9.

Magisus commented on lib/r10k/module/tarball.rb:
```diff
@@ -0,0 +1,95 @@
+  end
+
+  # Return the module's cachedir. Subclasses that implement a cache
+  # will override this to return a real directory location.
```
I think this comment got copied from `Module::Base` and isn't accurate here.

reidmv replied:
Totally. Fixed in c95c9c9.

Magisus commented on spec/unit/module/tarball_spec.rb:
```diff
@@ -0,0 +1,71 @@
+  end
+
+  describe "setting the owner and name" do
+    describe "with a title of 'mock-tarball'" do
```
I don't see this title coming from anywhere, does this `describe` need updating?

reidmv replied:
Indeed it does. Fixed in c95c9c9.

Magisus commented on spec/unit/module/tarball_spec.rb:
```diff
@@ -0,0 +1,71 @@
+  describe "determining the status" do
+    it "delegates to R10K::Tarball" do
+      mock_tarball = instance_double('R10K::Tarball', cache_valid?: true, insync?: true)
+      expect(subject).to receive(:tarball).twice.and_return instance_double('R10K::Tarball', cache_valid?: true, insync?: true)
```
Did you mean to use `mock_tarball` here as the thing to return?

reidmv replied:
Either that or don't declare a variable for this. I went that route and deleted the line in c95c9c9.

Magisus commented on spec/unit/module/tarball_spec.rb:
```diff
@@ -0,0 +1,71 @@
+    describe "version" do
+      context "when no version is given" do
+        subject { described_class.new('fixture-tarball', moduledir, base_params.reject { |k| k.eql?(:version) }) }
+        it "does not require a version" do
```
So, from a user's perspective, what does this do? I assume since these tarballs are downloaded from somewhere specific, they still would get a fixed version. Does this just allow them to not specify the checksum explicitly?

reidmv replied:
I added a mention of that to the docs in bc24f6b. Basically, if the user doesn't specify a checksum, the cache name will be based on the source URL and won't be invalidated until/unless the source URL changes, or a version string is supplied. With no version, whatever gets downloaded and cached is always considered valid and correct.

Magisus commented on lib/r10k/environment/tarball.rb:
```diff
@@ -0,0 +1,78 @@
+    when :absent, :mismatched
+      tarball.unpack(path.to_s)
+      # TODO: Remove untracked files. It's possible r10k's purge facility will
+      #       kinda sorta deal with this, but we shouldn't rely on it.
```
Given what you've done with `desired_contents` below, I think r10k's purging will take care of this, assuming environment-level purging is enabled (which it is in PE). Are you concerned about the case where it's disabled? Or was there some other edge case you had in mind?

reidmv replied:
I didn't have any specific edge cases in mind, it just felt unclean to leave it. But you're right, this is basically what r10k's purging is for. Updated comment to indicate recognition that old files could be left behind, and that we are relying on purging to come through and clean up after.

Magisus reviewed (commented):
I think I want to play with this a bit.

reidmv committed:
Incorporate feedback from review

Co-authored-by: Maggie Dreyer <maggie@puppet.com>

reidmv committed:
Add tarball module and environment type docs

Magisus commented on lib/r10k/environment/tarball.rb:
```diff
@@ -0,0 +1,78 @@
+  # @return [Array<String>]
+  def desired_contents
+    desired = []
+    desired += @tarball.paths.map { |entry| File.join(@full_path, entry) }
```
Discovered an interesting edge case here: if you build the tarball a bit wrong, even though it unpacks correctly, it might purge incorrectly. In my case, I created the tarball with `.`. So the paths inside it look like this:
```
"/Users/aileen/code/r10k/yaml-envs/tarball/.",
 "/Users/aileen/code/r10k/yaml-envs/tarball/./site-modules",
 "/Users/aileen/code/r10k/yaml-envs/tarball/./hiera.yaml",
...
```
But r10k's `current_contents` list looks like this:
```
"/Users/aileen/code/r10k/yaml-envs/tarball/site-modules",
 "/Users/aileen/code/r10k/yaml-envs/tarball/hiera.yaml",
...
```
So it purges everything except the `modules` directory (which is ignored by the `environment` purge level).

I realize this was my mistake, and I should get better at tarballs 😛 But it seems like something that could trip people up and cause a very confusing bug. Maybe we can add some normalization in `Tarball.paths`? Adding `Pathname.new(path).cleanpath` seemed to do the trick for me.

reidmv replied:
Ha! Excellent catch. Although I do have unbridled confidence that users of software would NEVER imperfectly construct a tarball 😉, this does seem like something we could fix.

reidmv replied:
Huh. I just snap-pushed what I thought would be the fix in 5a1711e, and then realized I was looking at a completely different part of the code. But, it might actually be a valid fix anyway? Don't have a test for it yet.

Magisus replied:
That's the spot I edited when fixing this locally 🤷‍♀️ Seemed to work for my bad tarball.

Magisus replied:
Were you working on a test for this? I think besides that, I'm happy to merge this.

reidmv replied:
Yeah, lemme add some thing for that. I might just re-construct the fixture tarball to include the `./` name components, I think that'd make it easy to validate. Will push something shortly.

Magisus replied:
Yeah that's what I was thinking 🙂

reidmv committed:
Use Pathname#cleanpath on Tarball paths

Users might construct a tarball with entries like this:

  ./
  ./foo.txt
  ./foo/bar/

r10k's purging facility needs these to be normalized to work correctly.
This commit makes the R10K::Tarball#paths method normalize the values to
e.g.

  foo.txt
  foo/bar

Co-authored-by: Maggie Dreyer <maggie@puppet.com>

Magisus reviewed (approved):
Retested this, seems to work well.

justinstoller commented on lib/r10k/module/tarball.rb:
```diff
@@ -0,0 +1,101 @@
+  def properties
+    {
+      :expected => @checksum,
+      :actual   => status,
```
This would be something like `:insync`, `:absent`, `:mismatch`, right? I think consumers of this (for existing modules) determine the status of insync/mismatch/absent from whether the actual equals the expected, or if it differs, and if it differs if the actual was nil, respectively. I think we want to continue that behavior.

reidmv replied:
Better? e783dcc6366d8ec6badbdc5a8c0d6d2b6ddc71ec

I'm not actually sure how/where/why #properties is used, or what expectations exist for this method. The YARDoc definition in R10K::Module::Base isn't super helpful on that front.

justinstoller replied:
They are collected to write the r10k-deploy.json, which is used by Impact Analysis.

reidmv committed:
Normalize tarball module properties return

When a module is insync, the module's `actual` and `expected` properties
should match. When the module is not in sync, they should differ.
